### PR TITLE
Write blog post on system dependencies in GitHub Actions continuous integration workflows

### PR DESCRIPTION
**Pull request by Bisaloo**
TODO:

- [x] Add examples of the two methods to deal with non-packaged system libraries with the [rbi R package](https://github.com/sbfnk/rbi)
- [x] Add links out to external documentation & other posts

I still have to go through it a couple of times but I think feedback at this stage would already be valuable.

### CODE REVIEW

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+      },
+      ">}}\n"
+    )
+  }
+)
+options(crayon.enabled = FALSE)
+
+```
+
+In a previous post, we discussed about a package dependency that goes slightly beyond the normal R package ecosystem dependency: R itself.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:39]:
```suggestion
In a [previous post](/2022/09/12/r-dependency/), we discussed about a package dependency that goes slightly beyond the normal R package ecosystem dependency: R itself.
```

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+Today, we step even further and discuss about dependencies that have nothing to do with R: system dependencies.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:40]:
```suggestion
Today, we step even further and discuss dependencies that have nothing to do with R: system dependencies.
```

I'd rephrase "nothing to do with R", because we still want to use them from R.

Comment (maelle) [reply content/post/2023-06-06-system-dependencies/index.Rmd:40]: I'd recommend adding examples of system dependencies and related packages right in the intro, to clarify the topic.

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+In particular, we are going to talk about system dependencies in the context of automated testing: is there anything extra to do when setting continuous integration for your package with system dependencies?
+How does it work behind the scenes?
+And how to work with edge cases?
+
+## Introduction: specifying system dependencies in R packages
+
+Before jumping right into the topic of continuous integration, let's take a moment to introduce, or remind you, how system dependencies are specified in R packages.
+You can directly jump to the next session if this concept is fresh in your memory.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:48]: I'd remove this, the reader doesn't need this permission slip. :smile_cat: 
```suggestion
```

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+Let's imagine we have a package depending on The official 'Writing R Extensions' guide states [^1]:
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:50]: incomplete sentence?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+[^1]: For R history fans, this has been the case [since R 1.7.0](https://github.com/r-devel/r-svn/blob/9c46956fd784c6985867aca069b926d774602928/doc/NEWS.1#L2348-L2350).
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:52]: add the year of that version?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+> Dependencies external to the R system should be listed in the 'SystemRequirements' field, possibly amplified in a separate README file.
+
+One important thing to note is that this field contains free text.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:56]: mention that as such, it's not used by R automatically at all?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+``` yaml
+SystemRequirements: ExternalSoftware 0.1
+```
+
+``` yaml
+SystemRequirements: lib-externalsoftware
+```
+
+## The general case: everything works automagically
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:71]: I once read mailchimp removed the word "automagically" from their style guide but I can't find the source again nor remember the reason. :sob:

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+If while reading the previous section, you could already sense the problems linked to the fact `SystemRequirements` is a free-text field, fret not!
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:73]: I'd add some sort of spoiler in the previous section ("free-text (yes, free-text...)") or so then

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+In the very large majority of cases, setting up continuous integration in an R package with system dependencies is exactly the same as with any other R package.
+
+Using, as often, the supercharged usethis package, you can automatically create the relevant GitHub Actions workflow file in your project [^2]:
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:76]: It'd be worth defining GitHub actions and introducing earlier that this will be the main example of CI in this post?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+[^2]: Alternatively, if you're not using usethis, you can manually copy-paste the relevant GitHub Actions workflow file from the `examples` of the `r-lib/actions` project.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:78]: link to it?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+          upload-snapshots: true
+```
+
+You may notice there is no explicit mention of system dependencies in this file.
+Yet, if we use this workflow in an R package with system dependencies, everything will work out-of-the-box in most cases.
+So, when are system dependencies installed?
+And how the workflow does even know which dependencies to install since the `SystemRequirements` is free text that may not correspond to the exact name of a library?
+
+The magic happens in the `r-lib/actions/setup-r-dependencies` step.
+If you want to learn about it, you can read the source code of this step.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:144]: link?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+It is mostly written in R but it contains a lot of bells and whistles to handle messaging within the GitHub Actions context and as such, it would be too long to go through it line by line in this post.
+However, at a glance, you can notice many mentions of the [pak R package](https://pak.r-lib.org/).
+
+If it's the first you're hearing about the pak package, we strongly recommend we go through the [list of the most important pak features](https://pak.r-lib.org/reference/features.html).
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:148]:
```suggestion
If it's the first time you're hearing about the pak package, we strongly recommend we go through the [list of pak features](https://pak.r-lib.org/reference/features.html).
```

This link is for ALL features?
Maybe mention the highlight `pak::pak()`?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+It is packed with many very powerful feature.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:149]: maybe embrace the pun
```suggestion
It is paked, hum packed with many very powerful features.
```

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+The specific feature we're interested in here is the automatic install of system dependencies via [`pak::pkg_sysreqs()`](https://pak.r-lib.org/reference/local_system_requirements.html), which in turn uses `pkgdepends::sysreqs_install_plan()`.
+
+We now understand more precisely where the magic happens but it still doesn't explain how pak is able to know which precise piece of software to install from the free text `SystemRequirements` field.
+As often when you want to increase your understanding, it is helpful to [read the source](https://blog.r-hub.io/2019/05/14/read-the-source/).
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:153]:
```suggestion
As often when you want to increase your understanding, it is helpful to [read the source](/2019/05/14/read-the-source/).
```

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+While browsing pkgdepends source code, we see a call to <https://github.com/r-hub/r-system-requirements>.
+
+This repository contains a set of [rules](https://github.com/rstudio/r-system-requirements/tree/main/rules) as json files which match unformatted software name via regular expressions to the exact libraries for each major operating system.
+Let's walk through an example together:
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:157]: if you use an example in the intro, then use it here again.

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+
+```{r}
+stringr::str_match("cudaa", c("\\bnvcc\\b", "\\bcuda\\b"))
+```
+
+This test confirms that the `SystemRequirements` field contents are not recognized by the regular expression.
+Depending on the case, the best course of action might be to:
+
+-   either edit the contents of `SystemRequirements` so that it's picked up by the regular expression
+-   or submit a pull request to `r-hub/r-system-requirements` if you believe the regular expression is too restrictive and should be updated
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:202]: link to an example PR?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+
+Note however that the first option is likely always the simplest as it doesn't impact all the rest of the ecosystem (which is why `r-system-requirements` maintainers might be reluctant to relax a regular expression) and it is often something directly in your control, rather than a third-party.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:204]: also it might take more time to wait for the PR to be reviewed

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+
+### Install system dependencies manually
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:206]:
```suggestion
### Install system dependencies "manually"
```

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+
+      - uses: r-lib/actions/check-r-package@v2
+```
+
+#### Using a Docker image in GitHub Actions
+
+Alternatively, you can do the manual installation in a Docker image and use this image in your GitHub Actions workflow.
+This is a particularly good solution if there is already a public Docker image or you already wrote a `DOCKERFILE` for your own local development purposes.
+If you use a public image, you can follow the steps in the official documentation to integrate it to your workflow.
+If you use a `DOCKERFILE`, you can follow [the answers to this stackoverflow question](https://stackoverflow.com/q/61154750/4439357) (in a nutshell, use `docker compose` in your job or publish the image first and then follow the official documentation).
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:250]: could you link to some basic tutorial as well? and to an example in a GitHub Action workflow?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+
+## Conclusion
+
+In this post, we have quickly reviewed how to specify system requirements for R package, how this seemingly innocent task requires a very complex infrastructure so that it can be understood by automated tools and that your dependencies are smoothly installed in a single command.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:254]:
```suggestion
In this post, we have provided an overview of how to specify system requirements for R package, how this seemingly innocent task requires a very complex infrastructure so that it can be understood by automated tools and that your dependencies are smoothly installed in a single command.
```

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+We also gave some pointers on what to do if you're in one of the rare cases where the automated tools don't or can't work.
+
+One final note on this topic is that there might be a move from CRAN to start requiring more standardization in the `SystemRequirements` field.
+One R package developer has reported being asked to change "Java (\>= 8)" to "Java JRE 8 or higher".
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:258]: is there a link to this report?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+In all cases, it is probably good practice to do this on your own and check what other R packages with similar system dependencies are writing in `SystemRequirements`.
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:259]: I'd actually move this to the part where you present the field.

**Review by maelle — COMMENTED** [review]
Thanks so much! This post will be very useful!

I added a few comments.

**Comment by gaborcsardi** [comment]
Thanks, looks great! Are you aware of the new stuff in pak, coming very soon? I.e. https://pak.r-lib.org/dev/reference/sysreqs.html ?

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 2 above)
Comment (Bisaloo) [inline content/post/2023-06-06-system-dependencies/index.Rmd:40]:
```suggestion
Today, we step even further and discuss dependencies outside of R: system dependencies.
```

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 4 above)
Comment (Bisaloo) [inline content/post/2023-06-06-system-dependencies/index.Rmd:50]:
```suggestion
The official 'Writing R Extensions' guide states [^1]:
```

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 7 above)
Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd:70]: The only reason I can imagine is that it might be unclear for non-native speakers. I can maybe add a link to https://english.stackexchange.com/questions/26618/what-is-the-purpose-of-using-the-word-automagically-when-we-already-have-auto behind the word?

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 21 above)
Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd:257]: Unfortunately, it was a thread in the grrr slack workspace so nothing public.

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 12 above)
Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd:147]: I'm not sure. The vignette is named "A list of the most important pak features"

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 16 above)
Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd:205]: @gaborcsardi, could you confirm that `r-hub/r-system-requirements` is really the repo where PR should be submitted? Or does it make more sense to submit it to the upstream `rstudio/r-system-requirements`?

Comment (gaborcsardi) [reply content/post/2023-06-06-system-dependencies/index.Rmd:205]: It should be submitted upstream at `rstudio/r-system-requirements`.

Comment (gaborcsardi) [reply content/post/2023-06-06-system-dependencies/index.Rmd:205]: @Bisaloo In fact, I wouldn't even mention `r-hub/r-system-requirements`, only the upstream repo.

Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd:205]: Great, thanks for your fast answer! This also means that changes will take time to propagate even when merged upstream? Or do you automatically regularly update the fork?

Comment (gaborcsardi) [reply content/post/2023-06-06-system-dependencies/index.Rmd:205]: Changes do take time, nothing uses that repo "live", so PPM changes take time, and pak changes take time as well, and yes, the latter are manual. 
This is to avoid having an upstream breaking change, e.g. a schema change, breaking pak.

**Comment by Bisaloo** [comment]
> Thanks, looks great! Are you aware of the new stuff in pak, coming very soon? I.e. [pak.r-lib.org/dev/reference/sysreqs.html](https://pak.r-lib.org/dev/reference/sysreqs.html) ?

Thanks, I was aware of it by looking at the source code but didn't notice the vignette. That might have been faster to understand how it works :sweat_smile:.

I have added a note about this in [`f1c4e20` (#165)](https://github.com/r-hub/blog/pull/165/commits/f1c4e20e9f55de92e594b35c48488a9c2d14e048). Please let me know if you'd rather see it mentioned somewhere else or in a different manner.

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 19 above)
Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd:255]: I will add a link to this example once it's finalized: https://github.com/sbfnk/rbi/pull/46 (shouldn't be long once I'm back from leave)

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 7 above)
Comment (maelle) [reply content/post/2023-06-06-system-dependencies/index.Rmd:76]: Or just take the risk to leave it as is, I feel that's better than adding a link. :thinking:

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 12 above)
Comment (maelle) [reply content/post/2023-06-06-system-dependencies/index.Rmd]: oh right :woman_facepalming:

**Review by maelle — COMMENTED** [review]
Thanks @Bisaloo!

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 21 above)
Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd:267]: This is the message in case you're curious: https://r-grrr.slack.com/archives/C9N6AESJG/p1682499615940129?thread_ts=1682495187.488019&cid=C9N6AESJG

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+
+If it's the first time you're hearing about the pak package, we strongly recommend we go through the [list of the most important pak features](https://pak.r-lib.org/reference/features.html).
+It is ~~paked~~ packed with many very powerful features.
+The specific feature we're interested in here is the automatic install of system dependencies via [`pak::pkg_sysreqs()`](https://pak.r-lib.org/reference/local_system_requirements.html), which in turn uses `pkgdepends::sysreqs_install_plan()`.
+
+We now understand more precisely where the magic happens but it still doesn't explain how pak is able to know which precise piece of software to install from the free text `SystemRequirements` field.
+As often when you want to increase your understanding, it is helpful to [read the source](/2019/05/14/read-the-source/).
+While browsing pkgdepends source code, we see a call to <https://github.com/r-hub/r-system-requirements>.
+
+This repository contains a set of [rules](https://github.com/rstudio/r-system-requirements/tree/main/rules) as json files which match unformatted software name via regular expressions to the exact libraries for each major operating system.
```
[the remark this answers is not in the document]
Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd:161]: The text is correct as it is right now as far as I can tell but there is a small unexplained gap:

- pkgdepends source code links to `r-hub/r-system-requirements`
- PRs should be sent `rstudio/r-system-requirements`

I don't explain the difference between the two and why I go from linking one to the other on the following line. I can maybe just add a note linking to [this thread](https://github.com/r-hub/blog/pull/165#discussion_r1280644182)?

[the remark this answers is not in the document]
Comment (maelle) [reply content/post/2023-06-06-system-dependencies/index.Rmd:161]: @gaborcsardi can you help?

File: content/post/2023-06-06-system-dependencies/index.Rmd [hunk 24]
```diff
@@ -0,0 +1,295 @@
...
+
+You can again see [a real-life example in the rbi R package](https://github.com/sbfnk/rbi/pull/46/files).
+
+## Conclusion
+
+In this post, we have provided an overview of how to specify system requirements for R package, how this seemingly innocent task requires a very complex infrastructure so that it can be understood by automated tools and that your dependencies are smoothly installed in a single command.
+We also gave some pointers on what to do if you're in one of the rare cases where the automated tools don't or can't work.
+
+One final note on this topic is that there might be a move from CRAN to start requiring more standardization in the `SystemRequirements` field.
+One R package developer has reported being asked to change "Java (\>= 8)" to "Java JRE 8 or higher".
```
Comment (maelle) [inline content/post/2023-06-06-system-dependencies/index.Rmd:293]:
```suggestion
One R package developer has reported being asked to change "Java JRE 8 or higher" to "Java (\>= 8)".
```
If I read the message correctly, it's the other way round.

**Review by maelle — APPROVED** [review]
Looks good to me, thanks so much! We can merge once the link issue is solved, please ping me in Slack so I can merge!

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 23 above)
[the remark this answers is not in the document]
Comment (gaborcsardi) [reply content/post/2023-06-06-system-dependencies/index.Rmd:161]: The rstudio repo is the canonical truth, PRs should go there.

File: content/post/2023-06-06-system-dependencies/index.Rmd (hunk 24 above)
Comment (Bisaloo) [reply content/post/2023-06-06-system-dependencies/index.Rmd]: You're completely right, thanks! I've read this message so many times. I'm not sure how I manage to understand the opposite :facepalm:
